src/polydatum/middleware.py

```diff
@@ -0,0 +1,184 @@
+from typing import Tuple, Any, Dict, Optional, Callable
+
+from polydatum.services import Service
+from polydatum.context import DataAccessContext
+
+
+class PathSegment:
+    """
+    This structure represents a segment of an attribute path.
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name = None
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta = None
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+    def __eq__(self, other):
+        return all([
+            isinstance(self, PathSegment),
+            self.name == other.name,
+            self.meta == other.meta
+        ])
+
+
+class DalMethodRequest:
+    """
+    Mutable request state. Middleware can modify this.
+    """
+
+    def __init__(
+            self, ctx: DataAccessContext, path: Tuple[PathSegment, ...], args: Tuple[Any, ...], kwargs: Dict
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, path: Optional[Tuple[PathSegment]] = None):
+        self.path = path
+        super().__init__(f'Invalid dal method: {self.path}')
```

Comment:
`self.path` is a tuple of `PathSegment`, so the output here won't be pretty. Something like:

```
Invalid dal method: (PathSegment(name="foo"), PathSegment(name="bar"))
```

Ideally it would be:

```
Invalid dal method: foo.bar
```

The caller doesn't know the internal structure being used here and wouldn't know how to map it back to their actual code line.